proposals/unions.md

```diff
@@ -0,0 +1,322 @@
+
+Case types can be any type that converts to `object`, e.g., interfaces, type parameters, nullable types and other unions. It is fine for cases to overlap, and for unions to nest or be null.
+
+#### Lowering
+
+A union declaration is lowered to a struct declaration with
+
+* the same attributes, modifiers, name, type parameters and constraints,
+* implicit implementations of `IUnion` and `IUnion<T>`,
+* a `public object? Value { get; }` auto-property,
+* a public constructor for each case type,
+* any members in the union declaration's body.
+
+If any user-declared members conflict with generated members, only the user-declared members are kept.
+
+Example:
+
+``` c#
+public union Pet(Cat, Dog){ ... }
+```
+
+Is lowered to:
+
+``` c#
+public record struct Pet : IUnion, IUnion<Pet>
+{
+    public Pet(Cat value) => Value = value;
+    public Pet(Dog value) => Value = value;
+    
+    public object? Value { get; }
+    
+    public static bool TryCreate(object? value, out Pet union)
+    {
+        (var result, union) = value switch
+        {
+            Cat v => (true, new Pet(v)),
+            Dog v => (true, new Pet(v)),
+            null => (true, new Pet()),
+            _ => (false, default),
+        };
+        return result;
+    }
+    
+    ... // original body
+}
+```
+
+## Open questions
+[open]: #open-questions
+
+* Both the use of constructors in union conversions and the use of `TryGetValue(...)` in union pattern matching are specified to be lenient when multiple ones apply: They'll just pick one. This should not matter per the well-formedness rules, but are we comfortable with it?
+* The specification subtly relies on the implementation of the `IUnion.Value` property rather than any `Value` property found on the union type itself. This is meant to give greater flexibility for existing types (which may have their own `Value` property for other uses) to implement the pattern. But it is awkward, and inconsistent with how other members are found and used directly on the union type. Should we make a change? Some other options:
+    * Require union types to expose a public `Value` property.
+    * Prefer a public `Value` property if it exists, but fall back to the `IUnion.Value` implementation if not (similar to `GetEnumerator` rules).
+* If a union type is a class type, it's value might itself be null. What about null checks then? The `null` pattern has been co-opted to check the `Value` property, so how do you check that the union itself isn't null? This part of the design is clearly optimized around the expectation that a union type is a struct. Some options:
+    * Too bad. Use `==` for your null check instead of a pattern match.
+    * Let the `null` pattern (and implicit null check in other patterns) apply to both the union value and it's `Value` property: `u is null ==> u == null || u.Value == null`.
+    * Disallow classes from being union types!
```

Comment:
Does this issue still exist in some form with structs as union types, when the union type is wrapped in `Nullable<T>`?